src/adapter.rs

```diff
@@ -1163,7 +1163,21 @@ impl<'a> Adapter<'a> for RustdocAdapter<'a> {
                                             token.as_impl().expect("not an Impl token");
 
                                         if let Some(path) = &impl_token.trait_ {
-                                            if let Some(item) = item_index.get(&path.id) {
+                                            // When the implemented trait is from the same crate
+                                            // as its definition, the trait should be present
+                                            // in `item_index`. Otherwise, the
+                                            // `rustdoc_types::Trait` is not in this rustdoc. 
+                                            // The needed foreign traits for lints are hand-written 
+                                            // in `dummy_trait_items`.
+                                            // See the comments in `src/indexed_crate.rs`
+                                            // for details.
+                                            let found_item = item_index
+                                                .get(&path.id)
+                                                .or_else(|| current_crate.dummy_trait_items.get(&path.id))
+                                                .or_else(|| previous_crate.and_then(|crate_| {
+                                                    crate_.dummy_trait_items.get(&path.id)
+                                                }));
```

Comment:
ID collisions across rustdoc JSON can happen, so this is setting us up for some _really bad_ debugging time in the future.

Never do the "try your luck" ID matching. Always explicitly figure out whether the ID is in the baseline or the current crate, and then look up the appropriate one.

Reply:
Done in 1f8530201a6544073f061823d91eb66d630914bf.